ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/task/GetHeadersFromPeerTask.java

```diff
@@ -0,0 +1,167 @@
+
+public class GetHeadersFromPeerTask implements PeerTask<List<BlockHeader>> {
+  private final long blockNumber;
+  private final Hash blockHash;
+  private final int maxHeaders;
+  private final int skip;
+  private final Direction direction;
+  private final int maximumRetriesAgainstDifferentPeers;
+  private final ProtocolSchedule protocolSchedule;
+  private final long requiredBlockchainHeight;
+
+  public GetHeadersFromPeerTask(
+      final long blockNumber,
+      final int maxHeaders,
+      final int skip,
+      final Direction direction,
+      final ProtocolSchedule protocolSchedule) {
+    this(blockNumber, maxHeaders, skip, direction, 5, protocolSchedule);
+  }
+
+  public GetHeadersFromPeerTask(
+      final long blockNumber,
+      final int maxHeaders,
+      final int skip,
+      final Direction direction,
+      final int maximumRetriesAgainstDifferentPeers,
+      final ProtocolSchedule protocolSchedule) {
+    this(
+        null,
+        blockNumber,
+        maxHeaders,
+        skip,
+        direction,
+        maximumRetriesAgainstDifferentPeers,
+        protocolSchedule);
+  }
+
+  public GetHeadersFromPeerTask(
+      final Hash blockHash,
+      final long blockNumber,
+      final int maxHeaders,
+      final int skip,
+      final Direction direction,
+      final ProtocolSchedule protocolSchedule) {
+    this(blockHash, blockNumber, maxHeaders, skip, direction, 5, protocolSchedule);
+  }
+
+  public GetHeadersFromPeerTask(
+      final Hash blockHash,
+      final long blockNumber,
+      final int maxHeaders,
+      final int skip,
+      final Direction direction,
+      final int maximumRetriesAgainstDifferentPeers,
+      final ProtocolSchedule protocolSchedule) {
+    this.blockHash = blockHash;
+    this.blockNumber = blockNumber;
+    this.maxHeaders = maxHeaders;
+    this.skip = skip;
+    this.direction = direction;
+    this.maximumRetriesAgainstDifferentPeers = maximumRetriesAgainstDifferentPeers;
+    this.protocolSchedule = protocolSchedule;
+
+    requiredBlockchainHeight =
+        direction == Direction.FORWARD
+            ? blockNumber + (long) (maxHeaders - 1) * (skip + 1)
+            : blockNumber;
+  }
+
+  @Override
+  public SubProtocol getSubProtocol() {
+    return EthProtocol.get();
+  }
+
+  @Override
+  public MessageData getRequestMessage() {
+    if (blockHash != null) {
+      return GetBlockHeadersMessage.create(
+          blockHash, maxHeaders, skip, direction == Direction.REVERSE);
+    } else {
+      return GetBlockHeadersMessage.create(
+          blockNumber, maxHeaders, skip, direction == Direction.REVERSE);
+    }
+  }
+
+  @Override
+  public List<BlockHeader> processResponse(final MessageData messageData)
+      throws InvalidPeerTaskResponseException {
+    if (messageData == null) {
+      throw new InvalidPeerTaskResponseException("Response MessageData is null");
+    }
+    return BlockHeadersMessage.readFrom(messageData).getHeaders(protocolSchedule);
```

Comment:
This should validate that the headers form a chain like the `AbstractGetHeadersFromPeerTask` task does

Reply:
Updated, including peer disconnection and chainstate updates when appropriate